sessions/return-oriented-programming/README.md

```diff
@@ -0,0 +1,810 @@
+# Return Oriented Programming
+
+## Recap - ASLR
+ASLR is not the only feature that prevents the compiler and the linker from
+solving some relocations before the binary is actually running. Shared libraries
+can also be combined in different ways, so the first time you actually know the
+address of a shared library is while the loader is running. The ASLR feature is
+orthogonal to this - the loader could choose to assign the addresses to
+libraries in a round-robin fashion, or could use ASLR to assign them randomly.
+
+Of course, we might be inclined to have the loader simply fix all relocations in
+the code section after it loaded the libraries, but this breaks the memory
+access protection of the `.text` section, which should only be **readable** and
+**executable**.
+
+## Solution - GOT and PLT
+In order to solve this issue, we need another level of indirection by which all
+accesses to symbols located in shared libraries will read the actual address
+from a table, called the **Global Offset Table (`.got`)**, at runtime. The one
+who populates this table is the loader. Note that this can work both for data
+accesses, as well as for function calls. However, function calls are actually
+using a small stub (i.e., a few instructions) stored in the **Procedure Linkage
+Table (`.plt`)**.
+
+The PLT is responsible of finding the shared library function address when it is
+first called (**lazy binding**), and writing it to a GOT entry. Note that the
+function pointers are stored in `.got.plt`). The following calls use the
+pre-resolved address. 
+
+Let's take a quick look at the code generated for a shared library call. We'll
+be using the binary compiled from the code below, which simply calls `puts()`.
+```c
+#include <stdio.h>
+
+int main(void)
+{
+	puts("Hello world!");
+	return 0;
+}
+```
+
+After compiling this code, let's look at the call to `puts()`:
+```
+$ objdump -D -j .text -M intel hello | grep puts
+80483e4:	e8 07 ff ff ff       	call   80482f0 <puts@plt>
+```
+
+If we look at the `.plt` section, we see that it starts at address `0x8049060`,

```

Comment:
Is `0x8049060` correct here?

Reply:
No. For some reason I kept the same addresses from the `.plt.sec` section. I'll change them.